java/src/main/native/src/row_conversion.cu

```diff
@@ -14,46 +14,86 @@
  * limitations under the License.
  */
 
+#include <algorithm>
+#include <cstdarg>
+#include <cstdint>
 #include <iostream>
+#include <iterator>
 #include <limits>
+#include <tuple>
+
+#include <cooperative_groups.h>
+#include <type_traits>
+
+#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 700
+#include <cuda/barrier>
+#endif
 
 #include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
 #include <cudf/detail/sequence.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/utilities/integer_utils.hpp>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/lists/lists_column_device_view.cuh>
 #include <cudf/scalar/scalar_factories.hpp>
 #include <cudf/table/table.hpp>
+#include <cudf/types.hpp>
 #include <cudf/utilities/bit.hpp>
 #include <cudf/utilities/error.hpp>
 #include <cudf/utilities/traits.hpp>
 #include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
 #include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+#include <thrust/binary_search.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/scan.h>
 
 #include "row_conversion.hpp"
 
+#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 700
+constexpr auto NUM_BLOCKS_PER_KERNEL_FROM_ROWS = 2;
+constexpr auto NUM_BLOCKS_PER_KERNEL_TO_ROWS = 2;
+constexpr auto NUM_BLOCKS_PER_KERNEL_LOADED = 2;
+constexpr auto NUM_VALIDITY_BLOCKS_PER_KERNEL = 8;
+constexpr auto NUM_VALIDITY_BLOCKS_PER_KERNEL_LOADED = 2;
+
+constexpr auto JCUDF_ROW_ALIGNMENT = 8;
+constexpr auto MAX_BATCH_SIZE = std::numeric_limits<cudf::size_type>::max();
+
+// needed to suppress warning about cuda::barrier
+#pragma nv_diag_suppress static_var_with_dynamic_init
+#endif
+
+using namespace cudf;
+using detail::make_device_uvector_async;
+using rmm::device_uvector;
 namespace cudf {
 namespace java {
+namespace detail {
 
```

Comment:
It might be worth centralizing various bookkeeping structs up here with a big block comment on what they all are and how they related to each other.  We've got a lot of similar terminology going on:  blocks, windows, batches, etc.  Seeing how they all relate in one place would be great.

Reply:
Done. Please comment on the block comment with suggestions if something isn't perfectly clear.